sycl/include/CL/sycl/buffer.hpp

```diff
@@ -351,6 +351,54 @@ class buffer {
         impl, reinterpretRange, OffsetInBytes, IsSubBuffer);
   }
 
+  // reinterpret() with no arguments. Can be specialized with one or two
+  // template arguments. four possible combinations two template specializers:
+  // R is same size as T. RD is same as dimensions.
+  // RD is one. To avoid ambiguity, R size is different or previous dimension is
+  // not same.
+
+  // single template specializers
+  // R is same size as T.  RD is assumed to be same as previous.
+  // R is not same as T  previous dimensions were 1.
+
+  template <typename ReinterpretT, int ReinterpretDim>
+  typename std::enable_if<
+      (sizeof(ReinterpretT) == sizeof(T)) && (dimensions == ReinterpretDim),
+      buffer<ReinterpretT, ReinterpretDim, AllocatorT>>::type
+  reinterpret() const {
+    return buffer<ReinterpretT, ReinterpretDim, AllocatorT>(
+        impl, get_range(), OffsetInBytes, IsSubBuffer);
+  }
+
+  template <typename ReinterpretT, int ReinterpretDim>
+  typename std::enable_if<
+      (ReinterpretDim == 1) && ((dimensions != ReinterpretDim) ||
+                                (sizeof(ReinterpretT) != sizeof(T))),
+      buffer<ReinterpretT, ReinterpretDim, AllocatorT>>::type
+  reinterpret() const {
+    long sz = get_size(); // TODO: switch to byte_size() once implemented
+    return buffer<ReinterpretT, ReinterpretDim, AllocatorT>(
+        impl, range<1>{sz / sizeof(ReinterpretT)}, OffsetInBytes, IsSubBuffer);
```

Comment:
I think you need to throw an exception if `sz` is not evenly divisible by `sizeof(ReinterpretT)`.  According to the spec:

> must throw an exception with the `errc::invalid` error code if the total size in bytes represented by the type and range of the reinterpreted SYCL buffer (or sub­-buffer) does not equal the total size in bytes represented by the type and range of this SYCL buffer(or sub-buffer).

Actually, I think the wording in the spec could also be clarified to say this more explicitly:

> must throw an exception with the `errc::invalid` error code if the size of this buffer is not evenly divisible by `sizeof(ReinterpretT)`.

However, I think the existing wording is equivalent (though confusing), so I think you should check for this case even if the spec is not clarified.

Reply:
https://github.com/KhronosGroup/SYCL-Docs/pull/121

Reply:
Done.